File: ios/hls/NitroPlayStreamRuntimeModule.swift
import Foundation
import React

@objc(NitroPlayStreamRuntime)
class NitroPlayStreamRuntime: NSObject, RCTBridgeModule {
  static func moduleName() -> String! { "NitroPlayStreamRuntime" }
  static func requiresMainQueueSetup() -> Bool { true }

  override init() {
    super.init()
  }

  @objc
  func start(_ port: NSNumber?) {
    HlsProxyRuntime.shared.start(port: port?.intValue)
  }

  @objc
  func stop() {
    HlsProxyRuntime.shared.stop()
  }

  @objc(getProxiedUrl:headers:)
  func getProxiedUrl(_ url: String, headers: NSDictionary?) -> String {
    HlsProxyRuntime.shared.getProxiedUrl(url: url, headers: headers as? [String: String])
  }

  @objc
  func prefetchFirstSegment(_ url: String, headers: NSDictionary?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    Task.detached {
      do {
        try await HlsProxyRuntime.shared.prefetchFirstSegment(url: url, headers: headers as? [String: String])
        resolver(true)
      } catch {
        rejecter("prefetch_error", error.localizedDescription, error)
      }
    }
  }

  @objc
  func getCacheStats(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    resolver(HlsProxyRuntime.shared.getCacheStats())
  }

  @objc
  func getStreamCacheStats(_ url: String, headers: NSDictionary?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    resolver(HlsProxyRuntime.shared.getStreamCacheStats(url: url, headers: headers as? [String: String]))
  }

  @objc
  func getThumbnailUrl(_ url: String, headers: NSDictionary?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    Task.detached {
      let result = await HlsProxyRuntime.shared.getThumbnailUrl(url: url, headers: headers as? [String: String])
      resolver(result)
    }
  }

  @objc
  func peekThumbnailUrl(_ url: String, headers: NSDictionary?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    resolver(HlsProxyRuntime.shared.peekThumbnailUrl(url: url, headers: headers as? [String: String]))
  }

  @objc
  func clearCache(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    HlsProxyRuntime.shared.clearCache()
    resolver(true)
  }

  @objc
  func clearPreview(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    HlsProxyRuntime.shared.clearPreview()
    resolver(true)
  }
}

File: android/src/main/java/com/nitroplay/hls/NitroPlayStreamRuntimeModule.kt
package com.nitroplay.hls

import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
class NitroPlayStreamRuntimeModule(private val reactContext: ReactApplicationContext)
    : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    init {
        reactContext.addLifecycleEventListener(this)
        HlsProxyRuntime.register(reactContext)
    }

    override fun getName(): String = "NitroPlayStreamRuntime"

    @ReactMethod
    fun start(port: Int) {
        HlsProxyRuntime.start(port)
    }

    @ReactMethod
    fun stop() {
        HlsProxyRuntime.stop()
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun getProxiedUrl(url: String, headers: ReadableMap?): String {
        return HlsProxyRuntime.getProxiedUrl(url, headers)
    }

    @ReactMethod
    fun prefetchFirstSegment(url: String, headers: ReadableMap?, promise: Promise) {
        HlsProxyRuntime.prefetchFirstSegment(url, headers, onComplete = {
            promise.resolve(true)
        }, onError = { err ->
            promise.reject("prefetch_error", err)
        })
    }

    @ReactMethod
    fun getCacheStats(promise: Promise) {
        promise.resolve(HlsProxyRuntime.getCacheStats())
    }

    @ReactMethod
    fun getStreamCacheStats(url: String, headers: ReadableMap?, promise: Promise) {
        promise.resolve(HlsProxyRuntime.getStreamCacheStats(url, HlsHeaderCodec.decode(headers)))
    }

    @ReactMethod
    fun clearCache(promise: Promise) {
        HlsProxyRuntime.clearCache()
        promise.resolve(true)
    }

    @ReactMethod
    fun clearPreview(promise: Promise) {
        HlsProxyRuntime.clearPreview()
        promise.resolve(true)
    }

    @ReactMethod
    fun getThumbnailUrl(url: String, headers: ReadableMap?, promise: Promise) {
        Thread {
            val result = HlsProxyRuntime.getThumbnailUrl(url, com.nitroplay.hls.HlsHeaderCodec.decode(headers))
            promise.resolve(result)
        }.start()
    }

    @ReactMethod
    fun peekThumbnailUrl(url: String, headers: ReadableMap?, promise: Promise) {
        Thread {
            val result = HlsProxyRuntime.peekThumbnailUrl(url, com.nitroplay.hls.HlsHeaderCodec.decode(headers))
            promise.resolve(result)
        }.start()
    }

    // LifecycleEventListener — self-heal on foreground return
    override fun onHostResume() {
        HlsProxyRuntime.onHostResume()
    }

    override fun onHostPause() {
        // NanoHTTPD can keep running in background
    }

    override fun onHostDestroy() {
        HlsProxyRuntime.onHostDestroy()
    }
}
